### PR TITLE
TechDocs: Feature/open stack swift publisher

### PR DESCRIPTION
**Pull request by erdoganoksuz**
## Hey, I just made a Pull Request!

We just added OpenStack Swift publisher for tech docs. 

Screen Shot 2021-02-26 at 10 44 15
Screen Shot 2021-02-26 at 10 45 05

#### :heavy_check_mark: Checklist

- [:heavy_check_mark: ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ :heavy_check_mark:] Added or updated documentation
- [:heavy_check_mark: ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

File: docs/features/techdocs/using-cloud-storage.md
Comment (OrkoHunter) [inline docs/features/techdocs/using-cloud-storage.md]: As we discussed on the group call, it's fine to start with adding these credentials in the config, but ideally they should follow the pattern of AWS/GCS publisher where we don't define anything in the config, and the client reads them from environment variables.

File: packages/techdocs-common/package.json
Comment (OrkoHunter) [inline packages/techdocs-common/package.json:59]: I did not know about this package. https://github.com/pkgcloud/pkgcloud#storage

Turns out we could just use `pkgcloud` as an abstraction over all cloud providers in TechDocs. But not critical as of now.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

**Review by OrkoHunter — COMMENTED** [review]
It's awesome that you got this working! Thank you for your contribution. 💯 

Left few small comments, will revisit later today - but looks good mostly!

File: docs/features/techdocs/using-cloud-storage.md
Comment (freben) [inline docs/features/techdocs/using-cloud-storage.md:286]: something went wrong in the text here

File: packages/techdocs-common/src/stages/publish/openStackSwift.ts
Comment (freben) [inline packages/techdocs-common/src/stages/publish/openStackSwift.ts:133]: please move the stream creation into the limiter below, so it doesn't create an unbounded number of streams at startup that aren't being read until later

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

File: packages/techdocs-common/src/stages/publish/openStackSwift.ts
Comment (freben) [inline packages/techdocs-common/src/stages/publish/openStackSwift.ts:252]: this one should probably be more robust with actual errors, right? so that they propagate as a rejection to the caller of hasDocsBeenGenerated

File: plugins/techdocs/config.d.ts
Comment (freben) [inline plugins/techdocs/config.d.ts:108]: secret!

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

File: app-config.yaml
Comment (mertcb) [reply app-config.yaml]: done that

File: packages/techdocs-common/__mocks__/pkgcloud.ts
Comment (mertcb) [reply packages/techdocs-common/__mocks__/pkgcloud.ts]: done that

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

File: app-config.yaml
Comment (adamdmharvey) [reply app-config.yaml]: I'll just add, this will keep growing and growing. I wonder at some point if we should change this to link to point here or something?

https://backstage.io/docs/features/techdocs/techdocs-overview#platforms-supported

**Review by adamdmharvey — COMMENTED** [review]
Not sure if I'm 100% on this, but this PR maybe should also change the supported file storage providers here?

https://backstage.io/docs/features/techdocs/techdocs-overview#platforms-supported

Unless there are more follow on needed to get Swift fully supported, but just wanted to throw that in!

File: app-config.yaml
Comment (freben) [reply app-config.yaml]: That also reminds me - this should totally have an "Integrations" entry in the docs as well! So people can see that it exists and how to configure it

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

File: packages/techdocs-common/package.json
Comment (erdoganoksuz) [reply packages/techdocs-common/package.json:59]: There is no official OpenStack SDK for node.js and we found this package in OpenStack wiki other packages outdated (last commit 4 years ago). This package maintainer not update last commit 17 months ago i think its not suitable for backstage. We plan to create updated OpenStack SDK they use ApplicationId and Secret for Auth.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

File: packages/techdocs-common/src/stages/publish/openStackSwift.ts
Comment (erdoganoksuz) [reply packages/techdocs-common/src/stages/publish/openStackSwift.ts]: @freben Hello I apologize for the late reply. I added warn log for this line. i don't understand what is your mean can you explain?

File: docs/features/techdocs/using-cloud-storage.md
Comment (Rugvip) [inline docs/features/techdocs/using-cloud-storage.md:371]: Think some of these could be left out from being specified using `$env` keys. I'm assuming it's very rare that the keystone version changes between deployments, so prolly just put the value straight in the config file instead

**Review by Rugvip — APPROVED** [review]
Thanks, looks good 👍 

Just a config example nit

Didn't dive deep into the implementation, will leave that to @backstage/techdocs-core

File: packages/techdocs-common/src/stages/publish/openStackSwift.ts
Comment (freben) [reply packages/techdocs-common/src/stages/publish/openStackSwift.ts]: I meant that you probably should not log at all. Instead do `if (err) { rej(err) }` so that the caller gets informed that there's an error because that's different from it not being present. Right?

Comment (erdoganoksuz) [reply packages/techdocs-common/src/stages/publish/openStackSwift.ts]: First of all, I check the `PublisherBase` class `hasDocsBeenGenerated` method should return a boolean instead of an error object. Here implementation; 
https://github.com/backstage/backstage/blob/4117d00f2177970a06f2ff4126cfb0a65cfdfcfd/packages/techdocs-common/src/stages/publish/types.ts#L75

When the requested file does not exist in the OpenStack Swift container `pkgcloud` returns a 404 error object. I think it's not an actual error because 404 meaning not found for this reason we call `res(false)`

Check google and s3 integrations they do the same thing, always return a boolean.
S3: https://github.com/backstage/backstage/blob/ae30b916df48cb554f171c04791bdc85be1a201a/packages/techdocs-common/src/stages/publish/awsS3.ts#L269

Google:
https://github.com/backstage/backstage/blob/ae30b916df48cb554f171c04791bdc85be1a201a/packages/techdocs-common/src/stages/publish/googleStorage.ts#L209

File: docs/features/techdocs/using-cloud-storage.md
Comment (erdoganoksuz) [reply docs/features/techdocs/using-cloud-storage.md:371]: We think the same thing `keystoneAuthVersion`, `domainId` and `domainName` optional values. You can see defaults here, https://github.com/erdoganoksuz/backstage/blob/d679b32f20f3771eced12488f244d27df7f4832d/packages/techdocs-common/src/stages/publish/openStackSwift.ts#L65

We don't know how to say these values optional in documentation if you give a suggestion we do this.

Maybe like that;
`keystoneAuthVersion?:`

Or
`keystoneAuthVersion: (optional)`

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4714-72wgh9zd7uxvrvi1p0voch5d7lyg9sko.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6038a769f285af6692994410

File: docs/features/techdocs/using-cloud-storage.md
Comment (Rugvip) [reply docs/features/techdocs/using-cloud-storage.md:371]: Yeah I think `keystoneAuthVersion: (optional)` works, and then perhaps only inline an example auth URL?

File: docs/features/techdocs/using-cloud-storage.md
Comment (Rugvip) [reply docs/features/techdocs/using-cloud-storage.md]: Is that because of the usage in the CLI? The backend also needs to be authenticated to read the data back though right? And we'd definitely want to use the config for that. Prolly makes sense to use this `fromConfig` in the backend but the constructor in the CLI in that case?

File: packages/techdocs-common/src/stages/publish/openStackSwift.ts
Comment (iamEAP) [inline packages/techdocs-common/src/stages/publish/openStackSwift.ts:229]: Kinda minor, but is there any reason not to pipe `this.storageClient.download()` directly through to `res`?  We're still buffering the whole file into memory before sending it to the client and it'd be nice to avoid the memory overhead.

The [GCS Publisher](https://github.com/backstage/backstage/blob/master/packages/techdocs-common/src/stages/publish/googleStorage.ts#L185) would be a good reference.

Comment (erdoganoksuz) [reply packages/techdocs-common/src/stages/publish/openStackSwift.ts:229]: We were inspired existing aws S3 publisher they do same thing. Next version we thing plan change OpenStack Swift package and some configuration. We can do this on the next version if it is no problem?

File: docs/features/techdocs/using-cloud-storage.md
Comment (OrkoHunter) [reply docs/features/techdocs/using-cloud-storage.md:371]: Agreed that we don't need to show the usage of `$env` for `keystoneAuthVersion`, `domainId` and `domainName`. Let's write inline docs like `keystoneAuthVersion: (optional, default: 'v3')`

`authUrl` can also be an inline string without `$env`.

File: docs/features/techdocs/using-cloud-storage.md
Comment (OrkoHunter) [reply docs/features/techdocs/using-cloud-storage.md]: > And we'd definitely want to use the config for that.

Not always. In case of S3 for example, if there is no credentials defined in the config, we create an instance of `aws-sdk` S3 client with empty parameters. The client is responsible for reading the environment variables on its own.

In case of OpenStack Swift, I am not sure how well the `pkgcloud` can read environment variables. Hence, I thought it's fine to explicitly define all these configs, but later on investigate to remove as much as possible, and leverage the environment variables that the relevant SDK client expects.


> Prolly makes sense to use this fromConfig in the backend but the constructor in the CLI in that case?

Yeah could be done.

**Review by OrkoHunter — APPROVED** [review]
Looks great!

@erdoganoksuz What are some follow-up issues/PRs in your mind about this? Some things that come to my mind

1. When a new release of `@backstage/techdocs-common` is made, bump the version in https://github.com/backstage/techdocs-cli and modify the `techdocs-cli publish` command to publish to OpenStack Swift.
2. Try to use Environment Variables (at least for username, password, etc.) and simplify defining credentials in the config. Investigate if the `pkgcloud` can read some environment variables on their own.
3. Stream files on download and upload like GCS/AWS publisher.

**Comment by erdoganoksuz** [comment]
> Looks great!
> 
> @erdoganoksuz What are some follow-up issues/PRs in your mind about this? Some things that come to my mind
> 
> 1. When a new release of `@backstage/techdocs-common` is made, bump the version in https://github.com/backstage/techdocs-cli and modify the `techdocs-cli publish` command to publish to OpenStack Swift.
> 2. Try to use Environment Variables (at least for username, password, etc.) and simplify defining credentials in the config. Investigate if the `pkgcloud` can read some environment variables on their own.
> 3. Stream files on download and upload like GCS/AWS publisher.

When I look at `pkgcloud` source code there are nothing environment variables. There is a way backstage we read environment variables ourselves?

 If not we don't find a good OpenStack SDK we must develop a new one.

We thought to contribute `pkgcloud` but they left alone the project. Last commit 17 months ago :sad:

**Comment by OrkoHunter** [comment]
> When I look at pkgcloud source code there are nothing environment variables. There is a way backstage we read environment variables ourselves?

Yeah you are write about that. Looking at https://github.com/pkgcloud/pkgcloud/blob/master/lib/pkgcloud/openstack/client.js, they do expect everything as config. We could create our own standard environment variables for for defining Open Stack username and password, but let's not go there yet. It's fine to have it in the config for now (and improve when more people start using OpenStack TechDocs Publisher.)

`pkgcloud` seems to be their recommended SDK package 😄 
https://wiki.openstack.org/wiki/SDKs#JavaScript

@erdoganoksuz Is this good to merge from your side?

**Comment by erdoganoksuz** [comment]
> > When I look at pkgcloud source code there are nothing environment variables. There is a way backstage we read environment variables ourselves?
> 
> Yeah you are right about that. Looking at https://github.com/pkgcloud/pkgcloud/blob/master/lib/pkgcloud/openstack/client.js, they do expect everything as config. We could create our own standard environment variables for defining Open Stack username and password, but let's not go there yet. It's fine to have it in the config for now (and improve when more people start using OpenStack TechDocs Publisher.)
> 
> `pkgcloud` seems to be their recommended SDK package 😄
> https://wiki.openstack.org/wiki/SDKs#JavaScript
> 
> @erdoganoksuz Is this good to merge from your side?

Yeah it's okay 👍

**Pull request merged**
